src/app/inbound.rs

```diff
@@ -262,129 +262,49 @@ pub mod rewrite_loopback_addr {
-    where
-        M: svc::Stack<Source>,
-    {
-        type Value = <Stack<M, B> as svc::Stack<Source>>::Value;
-        type Error = <Stack<M, B> as svc::Stack<Source>>::Error;
-        type Stack = Stack<M, B>;
-
-        fn bind(&self, inner: M) -> Self::Stack {
-            Stack {
-                inner,
-                _marker: PhantomData,
-            }
-        }
-    }
-
-    // === impl Stack ===
-
-    impl<M: Clone, B> Clone for Stack<M, B> {
-        fn clone(&self) -> Self {
-            Stack {
-                inner: self.inner.clone(),
-                _marker: PhantomData,
-            }
-        }
-    }
-
-    impl<M, B> svc::Stack<Source> for Stack<M, B>
-    where
-        M: svc::Stack<Source>,
-    {
-        type Value = svc::Either<Service<M::Value, B>, M::Value>;
-        type Error = M::Error;
-
-        fn make(&self, source: &Source) -> Result<Self::Value, Self::Error> {
-            let svc = self.inner.make(source)?;
-
+    pub fn layer() -> Layer<&'static str, Source, ReqHeader> {
+        add_header::request::layer(L5D_CLIENT_ID, |source: &Source| {
             if let Conditional::Some(ref id) = source.tls_peer {
                 match HeaderValue::from_str(id.as_ref()) {
                     Ok(value) => {
                         debug!("l5d-client-id enabled for {:?}", source);
-                        return Ok(svc::Either::A(Service {
-                            inner: svc,
-                            value,
-                            _marker: PhantomData,
-                        }));
+                        return Some(value);
                     }
                     Err(_err) => {
                         warn!("l5d-client-id identity header is invalid: {:?}", source);
                     }
-                }
+                };
             }
 
-            trace!("l5d-client-id not enabled for {:?}", source);
-            Ok(svc::Either::B(svc))
-        }
-    }
-
-    // === impl Service ===
-
-    impl<S: Clone, B> Clone for Service<S, B> {
-        fn clone(&self) -> Self {
-            Service {
-                inner: self.inner.clone(),
-                value: self.value.clone(),
-                _marker: PhantomData,
-            }
-        }
+            None
+        })
     }
+}
 
-    impl<S, B> svc::Service<http::Request<B>> for Service<S, B>
-    where
-        S: svc::Service<http::Request<B>>,
-    {
-        type Response = S::Response;
-        type Error = S::Error;
-        type Future = S::Future;
-
-        fn poll_ready(&mut self) -> Poll<(), Self::Error> {
-            self.inner.poll_ready()
-        }
-
-        fn call(&mut self, mut req: http::Request<B>) -> Self::Future {
-            req.headers_mut()
-                .insert(super::super::L5D_CLIENT_ID, self.value.clone());
-
-            self.inner.call(req)
-        }
+/// Adds `l5d-remote-ip` headers to http::Requests derived from the
+/// `remote` of a `Source`.
+pub mod remote_ip {
```

Comment:
It's pretty confusing that we have inbound::remote_ip and outbound::remote_ip and that they do subtle different things. Can we choose clearer names that differentiate the behavior?

`set_remote_ip_on_{req,rsp}`?

Reply:
The naming we choose is reflected here in [main.rs](https://github.com/linkerd/linkerd2-proxy/blob/0bd13492d55dbce44b44e64d53351c539c54c3eb/src/app/main.rs#L493)

Both `client_id` and `remote_ip` serve the same purpose for their respective header. If I change the naming on this, I'll change the naming on both.

I like `set_{header}_on_{req, rsp}`, so it should help clarify what exactly the line `.push(client_id::layer()` is doing when building the stack. It also just introduces another naming convention for these modules as headers are added (which may not be often!).

Reply:
I think it makes sense to clarify the naming in both cases.